### DISCUSSION TITLE
Workflows for Immutable objects

### DISCUSSION THREAD

**Discussion by toby-griffiths** [symfony/symfony#43199] [Q&A] [answered]
Hi there,

I'm just looking at the Workflow component for use in our projects, however we use immutable objects throughout our codebase, and in the current form Workflows would not work, as they don't return the modified object.

I'd like to start the discussion as to whether this might be possible without breaking the current implementation.  Having looked at the `apply()` code, neither `leave()` nor `enter()` have a return value, so I was wondering if there was a way to modify this to return the updated, immutable object, and then somehow returning the new, modified instance?

A few possible options (with no coment on how sane they are!)…

1. Add an `applyToImmutable()` method that returned the new instance?
2. Add the result to the `Marking` object?!?
3. Simply use the updated object on the resulting events that are triggers, and then just hook into those events, perhaps even using a decorator to provide a convenient way to access these?
4. Add a 4th `apply()` argument that is a callback that gets called with the updated object?

I realised that none of these are ideal, so if anyone has any better suggestions I'm all ears.

I'm happy to put this all together into a PR to add this support if a workable solutiuon can be come up with first.

Would this be better raised as an issue?

**Comment by lyrixx** [comment, marked as answer]
You may wrap your immutable object in an adapter to make it work. I guess it should work fine like this?

**Reply by toby-griffiths** [reply]
Ah yes.  Good idea.  Thanks.
